core/src/main/scala/akka/persistence/r2dbc/internal/R2dbcExecutor.scala

```diff
@@ -301,26 +323,82 @@ class R2dbcExecutor(val connectionFactory: ConnectionFactory, log: Logger, logDb
         result.failed.foreach { exc =>
           log.debug2("{} - DB call failed: {}", logPrefix, exc)
           // auto-commit so nothing to rollback
-          connection.close().asFutureDone()
+          connection.close().asFutureDone().andThen { case _ => timeoutTask.cancel() }
         }
 
         result.flatMap { r =>
-          connection.close().asFutureDone().map { _ =>
-            val durationMicros = durationInMicros(startTime)
-            if (durationMicros >= logDbCallsExceedingMicros)
-              log.infoN("{} - DB call completed [{}] in [{}] µs", logPrefix, r, durationMicros)
-            r
-          }
+          connection
+            .close()
+            .asFutureDone()
+            .map { _ =>
+              val durationMicros = durationInMicros(startTime)
+              if (durationMicros >= logDbCallsExceedingMicros)
+                log.infoN("{} - DB call completed [{}] in [{}] µs", logPrefix, r, durationMicros)
+              r
+            }
+            .andThen { case _ => timeoutTask.cancel() }
         }
       }
     }
   }
 
   private def commitAndClose(connection: Connection): Future[Done] = {
-    connection.commitTransaction().asFutureDone().andThen { case _ => connection.close().asFutureDone() }
+    connection.commitTransaction().asFutureDone().andThen { case _ =>
+      try connection.close().asFutureDone()
+      catch ignoreConnectionClosedException
+    }
   }
 
   private def rollbackAndClose(connection: Connection): Future[Done] = {
-    connection.rollbackTransaction().asFutureDone().andThen { case _ => connection.close().asFutureDone() }
+    try connection
+      .rollbackTransaction()
+      .asFutureDone()
+      .andThen { case _ =>
+        try connection.close().asFutureDone()
+        catch ignoreConnectionClosedException
+
+      } catch ignoreConnectionClosedException
+  }
+
+  private def closeAfterTimeout(connection: Connection): Unit = {
+    import scala.concurrent.duration._
+    val timeout = 5.seconds
+
+    log.warn("Timeout after [{}], will try to close connection", timeout)
+
+    // PooledConnection will automatically try to rollback if connection is in transaction, but in case
+    // that also times out we will as a last resort try to close the underlying connection
+
+    val closeWrappedTask = connection match {
+      case wrapped: Wrapped[Connection] @unchecked =>
+        Some(
+          system.scheduler.scheduleOnce(
+            timeout,
+            { () =>
+              log.info("Timeout of closing connection (including rollback), will try to close underlying connection")
+              wrapped.close.asFutureDone().onComplete {
```

Comment:
Unwrap to get underlying connection?

```suggestion
              wrapped.unwrap.close.asFutureDone().onComplete {
```